### PR TITLE
TF-TRT Improve reshape op converter

### PR DESCRIPTION
**Pull request by tfeher**
This PR improves the Reshape op converter to handle explicit batch and dynamic shape input data.

Since TRT6, IShuffleLayer has an option to accept the shape parameter as an [input tensor](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_shuffle_layer.html#a0ab2ab37e4377c9c317c1d869908efbc). If the shape input tensor corresponds to an input tensor of  the TRTEngineOp, then we need to [specify the input shape values](https://github.com/tfeher/tensorflow/blob/318b5c416db756432125af9d25446bdc3db5ee1f/tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.cc#L127-L134) for the engine. Additionally we need to specify optimization profile parameters for the shape values. The necessary changes were implemented in TrtShapeOptimizationProfile, TRTEngineOp, trt_engine_utils.

The unit tests have only a single op, and therefore handling shape input values are required to test
reshape with tensor input shape. In normal network it is less likely to have shape input tensors. 

Additionally the handling of scalar input/output shapes are improved.

C++ test for serialization / deserialization of input shapes and shape values added.

Note: TensorRT has a function  [INetworkDefinition::markOutputForShapes](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_network_definition.html#a1f0fff6e9c2469e31ac3be82dd3f0f28) which could enable the output tensor's value to be computed by IExecutionContext::getShapeBinding. This is not used in this PR, but it might be useful for shape inference.

### CODE REVIEW

**Comment by gbaned** [comment]
@tfeher This PR is in draft, any update on this? Please. Thanks!

**Comment by gbaned** [comment]
@tfeher This PR is in draft, any update on this? Please. Thanks!

**Comment by tfeher** [comment]
I have updated the PR, it is ready for review.

**Comment by gbaned** [comment]
@tfeher  Can you please check @bixia1's comments and keep us posted ? Thanks!

**Comment by gbaned** [comment]
@tfeher  Any update on this PR? Please. Thanks!

**Comment by tfeher** [comment]
There is an issue with the converter using TRT 7.1.3. I am working on a fix.

**Comment by gbaned** [comment]
@tfeher  Any update on this PR? and please resolve conflicts Thanks!

**Comment by tfeher** [comment]
The unit test of the converter fails in explicit batch mode with TRT 7.1.3 due to a bug in TRT. Waiting for feedback from the TRT team.

**Comment by gbaned** [comment]
@tfeher  Any update on this PR? and please resolve conflicts Thanks!

**Review by tfeher — COMMENTED** [review]
Sorry for the delay. I have addressed the first comments, and disable the converter path affected by a bug in TRT. @bixia1 we can continue with the review.

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h:191]: I can see it is very convenient to append the shape_values to the end of the execution tensor shapes. I have two main questions regarding this:
(1) When selecting a cuda engine: Does OptimizationProfileConfig::IncludesShapes really need to be applied to the shape_values that are actually used as input tensors for reshape?  We also apply the check of OptimizationProfileConfig::IncludesShapes  to the shape values that are NOT used as input tensors for reshape, is that a problem? I tend to think it is a problem, and we will need to fix TrtShapeOptimizationProfile::GetProfileNumber and add test case for this.
(2) Is it possible to delay the copy from device to host of the shape_value in routine AppendShapeValues to TrtShapeOptimizationProfile::ConfigureBuilder? This can avoid unnecessary memcpy.

**Review by bixia1 — COMMENTED** [review]
Thanks for the work!  I haven't finished reviewing everything, but I think I should send out the comments I have at this point.

**Comment by gbaned** [comment]
@tfeher  Can you please check @bixia1's comments and keep us posted ? Thanks!

**Comment by gbaned** [comment]
@tfeher Any update on this PR? Please. Thanks!

File: tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.cc
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.cc:101]: This is possible, but we need actually two inputs: the vector of Tensor* and a vector of input names. I had such an implementation earlier, which I rejected to keep `TRTEngineOp::ExecuteTrtEngine` shorter. I can bring this back.

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the review and the useful suggestion. There are two issues still open, I am working on them.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (DEKHTIARJonathan) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:2283]: Will need to update with: `OpConverter_FP32_Test`

**Review by DEKHTIARJonathan — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@tfeher  Can you please check @DEKHTIARJonathan's comments and keep us posted ? Thanks!

**Comment by gbaned** [comment]
@tfeher Any update on this PR? Please. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by tfeher** [comment]
This PR depends on #45588, I will update here once #45588 is finalized.

File: tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.cc
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.cc:102]: I was thinking about the proposed change. I am not convinced if it simplifies things. Let me know if I should go forward with this:

1. For the inputs it could work the following way: 
- we need a helper to construct the vector of tensors from `OpKernelContext` (since the `params_` field is private),
- another helper to map the unit test `DataVec -> gtl::InlinedVector<TensorValue,4>`.

2. The outputs need either to be allocated using `OpKernelContext`, or just reshaped in case of the unit tests. One would need two more helpers that return the vector of tensors: 
- create a helper functions doing the allocations for `OpKernelContext`
- another one for the reshape operations during unit test with `DataVec`.

My preferred solution would be to make the converter tests a subclass of OpTestBase, and run the tests as a TF op test. But that would be outside the scope of this PR.

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h]: Update: I have modified shape value handling the following way:
- `TrtShapeOptimizationProfile::CollectShapeValues` will store the shape values into the `actual_shape_values_` array. During profile collection, we maintain two vectors: `input_shapes_` and `input_shape_values_`. 
- After profile collection, in `InitProfiles` we concatenate the the vectors from `input_shapes_` and `input_shape_values_`. For both type of inputs we need to perform the same operations when we define the actual profiles, that is the motivation to treat them as a single `std::vector<nvinfer1::Dims>`.

1.b An `OptimizationProfileConfig` contains `2*n_input` values in its `min`, `max`, `opt` members, where `min[n_input + k]` ( 0 <= k < n_input) can describe a shape value input if. If input `k` is not a shape tensor, then the corresponding values will be initialized as `min = opt = max = nvinfer1::Dims{0, {}}`, This is because `CollectShapeValues` zero inits components of `actual_shape_values_` that do not correspond a input shape tensor. Such values will always pass the `IncludesShape` check called by GetProfileNumber.

1.c Once the network is initialized (or engine is loaded from file) we use `SetShapeTensorMask` to query which tensors are actually shape tensors. After this step `CollectShapeValues` will only copy data for inputs that are marked as input shape binding of the engine.

2. Synchronization to ensure that shape values are copied to host: currently it is done in `CollectShapeValues`. This makes the control flow clear. Alternatively we could postpone the sync call to `AppendShapeValues` and `SetInputShapeBinding`.

**Comment by tfeher** [comment]
@bixia1 this is basically ready for review. But this PR contains the changeset from #45588 which makes it harder to read until #45588 is merged. I will rebase the PR once #45588 is merged.

I have improved the shape value handling as [described here](https://github.com/tensorflow/tensorflow/pull/40545#discussion_r572285311).

**Comment by tfeher** [comment]
@bixia1 I have rebased the PR, it is ready for review.

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:690]: CollectShapeValues preprares actual_shape_values for AddShape. From this code block, it is possible that AddShape won't be called after CollectShapeValues. It is not straightforward to understand.
Can we combine CollectShapeValues and AddShape and remove actual_shape_values?

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h:226]: Would this be better?
// The tensor values for the inputs with shape-tensor-compatible shapes. 
std::vector<std::vector<absl::optional<nvinfer1::dims>>>> maybe_shape_tensor_values;

**Review by bixia1 — COMMENTED** [review]
Thanks for your work! I finish reading most of the code but haven't looked at the tests yet.  I would like to send out my comments and review the test later.

**Comment by gbaned** [comment]
@tfeher Can you please check @bixia1's comments and keep us posted ? Thanks!

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the review! I have addressed the issues.

**Comment by tfeher** [comment]
I have rebased the PR to resolve the merge conflicts.

**Comment by tfeher** [comment]
I have noticed a problem with checking the shape value profiles during engine lookup. I am working on a fix.

**Comment by tfeher** [comment]
Fixed issue with shape value handling during profile selection, and added the missing datavec.h file. In a subsequent commit I will add one more python test to check shape value handling.

**Review by bixia1 — COMMENTED** [review]
Looks like you are still working on the PR. I am sending our the comment I have so far, before your  last commit.

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the additional comments. I have corrected the typos. We can continue with the review.

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_resource_ops_test.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_resource_ops_test.cc:129]: We can only allow this for TRT 6+, right?

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_resource_ops_test.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_resource_ops_test.cc:364]: Sorry, I don't understand the second part of this comment "used only order matters". Can you explain?
Should this comment be moved to before shape_values?

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the review, I have fixed the issues.

File: tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.cc
Comment (bixia1) [reply tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.cc:104]: I think making the converter tests run as OpTestBase is a good idea.

**Review by bixia1 — APPROVED** [review]
Approving this so that I can get it to google and test it.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:760]: do you mean if (dims.nDims == 0) ?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:2774]: input_tensor is not used.

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:675]: Clang doesn't like this, we need to add parentheses 
s/!use_implicit_batch_ && has_dynamic_shape_input_
/(!use_implicit_batch_ && has_dynamic_shape_input_
)/

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h:87]: input_dim is not used.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1491]: Need to handle the return value with TF_RETURN_IF_ERROR

**Comment by bixia1** [comment]
I got three tests failed:
(1) tensorflow/python/compiler/tensorrt:gpu_combined_nms_test. message is like
  File "/build/work/fe0e3a7166e31378da05c1404a4fcd4fedb1/google3/runfiles/google3/third_party/tensorflow/python/compiler/tensorrt/test/tf_trt_integration_test_base.py", line 1107, in _Test
    self.RunTest(run_params)
  File "/build/work/fe0e3a7166e31378da05c1404a4fcd4fedb1/google3/runfiles/google3/third_party/tensorflow/python/compiler/tensorrt/test/tf_trt_integration_test_base.py", line 1030, in RunTest
    GraphState.INFERENCE)
  File "/build/work/fe0e3a7166e31378da05c1404a4fcd4fedb1/google3/runfiles/google3/third_party/tensorflow/python/compiler/tensorrt/test/tf_trt_integration_test_base.py", line 907, in _VerifyGraphDef
    graph_state)
  File "/build/work/fe0e3a7166e31378da05c1404a4fcd4fedb1/google3/runfiles/google3/third_party/tensorflow/python/compiler/tensorrt/test/tf_trt_integration_test_base.py", line 893, in _VerifyGraphDefV2
    [name for name in unexpected_names if name in all_op_names])
AssertionError: ['score_threshold', 'max_output_size_per_class', 'iou_threshold', 'max_total_size'] has length of 4.
(2) tensorflow/python/compiler/tensorrt:gpu_neighboring_engine_test
and tensorflow/python/compiler/tensorrt:gpu_multi_connection_neighbor_engine_test
they both failed because of native segment execution happened,
  (0) Aborted: User disallowed engine native segment execution
	 [[{{node TRTEngineOp_0_1}}]]

**Comment by bixia1** [comment]
tensorflow/compiler/tf2tensorrt:convert_nodes_test_gpu also failed
[test.log](https://github.com/tensorflow/tensorflow/files/6098834/test.log)

**Review by bixia1 — CHANGES_REQUESTED** [review]

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the review. I have corrected the smaller issues. I am looking into the python test problems.

The C++ test seem to have a problem with ConvertConv3D. I could not reproduce that wint TRT 7.2.1, but I will have a look with other versions. 

There seem to be an additional issue with shape value handling in TRT6. I will investigate, and disable conversion in the affected versions.

**Comment by tfeher** [comment]
- I have disabled ConvertDynamicReshape before TRT 7.1.3, this way the reshape unit test works in all TRT versions that I have tested (5.1.5, 6.0.1, 7.1.3 and 7.2.1). 
- I could not reproduce the Conv3D test with TRT 7.1.3.4.
- I confirm the problems with the two neighbor engine tests. I will continue to investigate those.

**Review by bixia1 — APPROVED** [review]
Per offline conversation, approving this for testing inside google while Tamas is still working on the python tests.

**Comment by bixia1** [comment]
I found the reason for the Conv3D failure. [This line of code ](https://github.com/tensorflow/tensorflow/blob/8cb7f3a462db810d70521fab26eac8e2fed9e2b4/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L3428)is accessing garbage value when c_index == 4, and for implicit batch mode.
In case you need a test to verify it, [here is one](https://github.com/tensorflow/tensorflow/blob/8cb7f3a462db810d70521fab26eac8e2fed9e2b4/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L5018-L5032).
The problem is actually not introduced by this PR, but [this one](https://github.com/tensorflow/tensorflow/pull/46940).

**Review by bixia1 — APPROVED** [review]
Approving this so that I can get it to google and test it while Tamas is still working on one failing test.

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc]: Fixed.

**Review by tfeher — COMMENTED** [review]
The NMS Python test error was caused by a change in how scalars are represented in this PR. I have updated the converter, the test passes now.

I have also fixed the two smaller issues.

**Review by bixia1 — APPROVED** [review]

**Pull request merged**
